docs/content/en/docs/Advanced/bundles.md

```diff
@@ -0,0 +1,149 @@
+
+To apply a bundle before Kubernetes starts, you can include it in your config like this:
+
+```yaml
+#cloud-config
+
+bundles:
+- targets:
+  - run://<image>
+```
+
+Replace `<image>` with the URL or path to the bundle image. The prefix (e.g. `run://`) indicates the type of bundle being used.
+
+To install a bundle after installation instead (for those bundles that explicitly supports that), use the following:
+
+```yaml
+#cloud-config
+install:
+ bundles:
+ - targets:
+   - run://<image>
+```
+
+One of the benefits of using bundles is that they can also extend the cloud-config keywords available during deployment. This means that by adding bundles to your configuration file, you can add new blocks of configuration options and customize your system even further.
+
+A full config using a bundle from [community-bundles](https://github.com/kairos-io/community-bundles) that configures `metalLB` might look like this:
+
+```yaml
+#cloud-config
+
+hostname: kairoslab-{{ trunc 4 .MachineID }}
+users:
+- name: kairos
+  ssh_authorized_keys:
+  # Add your github user here!
+  - github:mudler
+
+k3s:
+  enable: true
+  args:
+  - --disable=servicelb
+
+# Specify the bundle to use
+bundles:
+- targets:
+  - run://quay.io/kairos/community-bundles:metallb_latest
+
+# Specify metallb settings
+metallb:
+  version: 0.13.7
+  address_pool: 192.168.1.10-192.168.1.20
+```
+
+## Bundle types
+
+Bundles can carry also binaries that can be overlayed in the rootfs, either while [building images](/docs/advanced/build) or with [Live layering](https://kairos.io/docs/advanced/livelayering/).
+
+
+Kairos supports three types of bundles:
+
+- *Container*: This type is a bare container that simply contains files that need to be copied to the system. It is useful for copying over configuration files, scripts, or any other static content that you want to include on your system.
+
+- *Run*: This type is also a bare container, but it comes with a script that can be run during the installation phase to add custom logic. This is useful for performing any necessary setup or configuration tasks that need to be done before the cluster is fully deployed.
+
+- *Package*: This type is a [luet](https://luet.io) package that will be installed in the system. It requires the luet repository definition in order to work. Luet packages are a powerful way to manage dependencies and install software on your system.
+
+Note: In the future, Kairos will also support a local type for use in airgap situations, where you can pre-add bundles to the image before deployment.
+
+It's important to note that bundles do not have any special meaning in terms of immutability. They install files over paths that are mutable in the system, as they are simply overlaid during the boot process. This means that you can use bundles to make changes to your system at any time, even after it has been deployed.
```

Comment:
So what happens if a bundle tries to write a file to an "immutable" (ro) path? If I had a bundle that provides a binary (e.g. `k9s`), would I be able to put it in `/usr/bin/`? 

I'm not sure the first sentence adds any information: `It's important to note that bundles do not have any special meaning in terms of immutability.`.  In a way, there is one thing we are breaking with bundles and that's the atomic installation of just one container image. Now the final installation may comprise of multiple images.

Maybe we should explain when bundles should be used over creating the final image directly. Some reasons I can think of:

- Use third-party bundles (consume other people's work)
- Re-use the same bundles in multiple (otherwise different) installations

Is there a way, to use bundles in an osbuilder Job so that a final image is generated that would be the equivalent of using the same bundles in the config?

Reply:
If a bundle writes to a **ro** path everything that would do would be discarded. It's responsibility of who writes the bundle to document its usage and implications.

Bundle are very generic ways to deliver specific tasks to agents,  that can span to dynamic configuration, adding statically files, but all of that needs to happen on paths that are mutable in the system.

osbuilder is privileged - as it can assembles new images can import directly what we call bundles in the root of the base image. but at that point bundles can just contain binaries - and I think should be treated separately or somehow push a layer with the dynamic configuration (our `run.sh`) as a hidden layer in the container image, but we do not have such a mechanism in place yet.